src/backend/joanie/core/models/products.py

```diff
@@ -720,20 +725,27 @@ def clean(self):
             )
             raise ValidationError({"__all__": [message]})
 
-        # The user should not be enrolled in another opened course run of the same course.
-        if (
-            self.created_on is None
-            and self.user.enrollments.filter(
+        if self.created_on is None:
+            # The user should not be enrolled in another opened course run of the same course.
+            if self.user.enrollments.filter(
                 course_run__course=self.course_run.course,
                 course_run__end__gte=timezone.now(),
                 is_active=True,
-            ).exists()
-        ):
-            message = _(
-                "You are already enrolled to an opened course run "
-                f'for the course "{self.course_run.course.title}".'
-            )
-            raise ValidationError({"user": [message]})
+            ).exists():
+                message = _(
+                    "You are already enrolled to an opened course run "
+                    f'for the course "{self.course_run.course.title}".'
+                )
+                raise ValidationError({"user": [message]})
+
+            # Enforce has_been_created_with_order to True if the enrollment relies on
+            # a not listed CourseRun
+            if not self.course_run.is_listed:
+                self.has_been_created_with_order = True
+            # Enforce has_been_created_with_order to False if the enrollment relies on
+            # a listed CourseRun linked to any product
+            elif not self.course_run.course.targeted_by_products.exists():
+                self.has_been_created_with_order = False
```

Comment:
I find this convoluted and think it would be clearer if we set the field by passing the order id from the front as discussed the other day. The code to check that the order is valid for enrollment will also be clearer. WDYT?

Reply:
Indeed clearer but coster in db queries and with any gain compare to the current check.